devicemanager: checkpoint: support pre-1.20 data

ffromani opened the pull request:
#### What type of PR is this?


/kind bug


#### What this PR does / why we need it:
The commit a8b8995ef241e93e9486d475126450f33f24ef4e changed the content of the data kubelet writes in the checkpoint.
Unfortunately, the checkpoint restore code was not updated,  so if we upgrade kubelet from pre-1.20 to 1.20+, the device manager cannot anymore restore its state correctly.
    
The only trace of this misbehaviour is this line in the  kubelet logs:
```
W0615 07:31:49.744770    4852 manager.go:244] Continue after failing to read checkpoint file. Device allocation info may NOT be up-to-date. Err: json: cannot unmarshal array into Go struct field PodDevicesEntry.Data.PodDeviceEntries.DeviceIDs of type checkpoint.DevicesPerNUMA
```
    
If we hit this bug, the device allocation info is indeed NOT up-to-date up until the device plugins register themselves again.
This can take up to few minutes, depending on the specific device plugin.
    
While the device manager state is inconsistent:
1. the kubelet will NOT update the device availability to zero, so the scheduler will send pods towards the inconsistent kubelet.
2. at pod admission time, the device manager allocation will not trigger, so pods will be admitted without devices actually being allocated to them.
    
To fix these issues, we add support to the device manager to read pre-1.20 checkpoint data. We retroactively call this format "v1".

#### Which issue(s) this PR fixes:
https://github.com/kubernetes/kubernetes/issues/102880

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
```release-note
Fixes a regression in 1.20 where the kubelet cannot read devicemanager checkpoint data persisted by a 1.19 kubelet
```

ffromani committed:
devicemanager: checkpoint: support pre-1.20 data

The commit a8b8995ef241e93e9486d475126450f33f24ef4e
changed the content of the data kubelet writes in the checkpoint.
Unfortunately, the checkpoint restore code was not updated,
so if we upgrade kubelet from pre-1.20 to 1.20+, the
device manager cannot anymore restore its state correctly.

The only trace of this misbehaviour is this line in the
kubelet logs:
```
W0615 07:31:49.744770    4852 manager.go:244] Continue after failing to read checkpoint file. Device allocation info may NOT be up-to-date. Err: json: cannot unmarshal array into Go struct field PodDevicesEntry.Data.PodDeviceEntries.DeviceIDs of type checkpoint.DevicesPerNUMA
```

If we hit this bug, the device allocation info is
indeed NOT up-to-date up until the device plugins register
themselves again. This can take up to few minutes, depending
on the specific device plugin.

While the device manager state is inconsistent:
1. the kubelet will NOT update the device availability to zero, so
   the scheduler will send pods towards the inconsistent kubelet.
2. at pod admission time, the device manager allocation will not
   trigger, so pods will be admitted without devices actually
   being allocated to them.

To fix these issues, we add support to the device manager to
read pre-1.20 checkpoint data. We retroactively call this
format "v1".

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/sig node
/triage accepted

fromanirh commented:
/remove-kind design
/remove-kind cleanup
/remove-kind documentation
/remove-kind feature

swatisehgal commented:
/cc

fromanirh commented:
pending items before to lift the WIP:
- doublecheck devices plugin which don't declare NUMA Affinity are assigned to NUMA cell 0
- doublecheck NUMA affinity is updated correctly on device registration after the device plugin registers itself

fromanirh commented:
> pending items before to lift the WIP:
> 
>     1. doublecheck devices plugin which don't declare NUMA Affinity are assigned to NUMA cell 0
> 
>     2. doublecheck NUMA affinity is updated correctly on device registration after the device plugin registers itself

2. covered by e2e tests. Doublechecking #1. I'm happy enough with the PR and it's reviewable.

fromanirh commented:
/priority important-longterm
not sure about the priority. When this bug triggers, it can lead to very bad and hard to debug behaviour.

fromanirh commented:
/priority important-soon
being an actual bug

fromanirh commented:
/remove-priority important-longterm

fromanirh commented:
> pending items before to lift the WIP:
> 
>     1. doublecheck devices plugin which don't declare NUMA Affinity are assigned to NUMA cell 0
> 
>     2. doublecheck NUMA affinity is updated correctly on device registration after the device plugin registers itself

regarding point 1:

1. when the plugins register themselves, the device information is rewritten with fresh data: https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/kubelet/cm/devicemanager/manager.go#L183 - this includes the topology information
2. the topology information is checked on-the-fly when needed: https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/kubelet/cm/devicemanager/manager.go#L758
3. if no topology is reported by a device, it is stored internally as affine to NUMA cell zero: https://github.com/kubernetes/kubernetes/blob/v1.21.2/pkg/kubelet/cm/devicemanager/manager.go#L915

So all the points are covered and this PR is fully ready

ehashman commented:
> Unfortunately, the checkpoint restore code was not updated, so if we upgrade kubelet from pre-1.20 to 1.20+, the device manager cannot anymore restore its state correctly.

Does this mean we need to backport this back to 1.20?

fromanirh commented:
> > Unfortunately, the checkpoint restore code was not updated, so if we upgrade kubelet from pre-1.20 to 1.20+, the device manager cannot anymore restore its state correctly.
> 
> Does this mean we need to backport this back to 1.20?

Yep, I think so:
```
$ git branch --contains a8b8995ef241e93e9486d475126450f33f24ef4e | grep release
  release-1.20
  release-1.21
```

The fix per se should apply cleanly to older branches, the e2e tests I added can be more challenging. But I'll take care of the backports should we decided to do them.

endocrimes commented on test/e2e_node/device_manager_test.go:
```diff
@@ -0,0 +1,370 @@
+		// this test wants to reproduce what happened in https://github.com/kubernetes/kubernetes/issues/102880
+		ginkgo.It("should be able to recover V1 (aka pre-1.20) checkpoint data and reject pods before device re-registration", func() {
+			if sriovdevCount, err := countSRIOVDevices(); err != nil || sriovdevCount == 0 {
+				e2eskipper.Skipf("this test is meant to run on a system with at least one configured VF from SRIOV device")
```
do we have a CI environment that can test this? (to avoid future regressions, and also out of my own curiosity as I learn more about that infra)

fromanirh replied:
Unfortunately we don't to date. In RH we have some machines with the SRIOV cards on which we run the tests, but for various reasons we can't expose them yet to internet. On k8s infra we don't have these devices.
For further context: most of the code require a device to be reported (so for example a fake device plugin would suffice) and sriov devices were just the easier to get running (compared to, random examples, gpus or fpgas or anything else)

endocrimes commented on pkg/kubelet/cm/devicemanager/checkpoint/checkpointv1.go:
```diff
@@ -0,0 +1,110 @@
+	Checksum checksum.Checksum
+}
+
+func NewV1(devEntries []PodDevicesEntryV1,
```
Can we add docstrings to the public parts of this file?  (I know we're not great at doing this consistently, but explicitly calling it out as a thing for backwards compatibility here would be useful for future readers)

endocrimes commented:
/retest

endocrimes commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -588,20 +588,45 @@ func (m *ManagerImpl) writeCheckpoint() error {
+		// one last try: maybe it's a old format checkpoint?
+		cp, err = m.restoreCheckpointV1()
+		if err != nil {
+			return err
```
I'm not sure we want to return the v1 error - Or at least not only the v1 error unless we also log the current version error

fromanirh replied:
logging the two errors looks like a nice idea to me

fromanirh replied:
actually decided to return the previous error (not the v1 error) because I couldn't find yet a good way to wrap two of them. I'm not very very fond of my current approach - even though I'd still call an improvement, so totally open for further discussion.

endocrimes replied:
That works I think, wrapping multi errors is always kinda uncomfortable

fromanirh commented:
/retest

fromanirh commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

endocrimes commented:
/lgtm

chenyw1990 commented:
/cc

ehashman commented:
/assign @klueska

ffromani committed:
node: e2e: add test for the checkpoint recovery

Add a e2e test to exercise the checkpoint recovery flow.
This means we need to actually create a old (V1, pre-1.20) checkpoint,
but if we do it only in the e2e test, it's still fine.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ehashman commented:
/assign @endocrimes

endocrimes reviewed (approved):
/lgtm

fromanirh commented:
begun cherry-pick process. Cherry-picking to 1.20 will not be trivial.
